segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/cache/DirectMemoryCache.java

```diff
@@ -0,0 +1,437 @@
+                DirectMemoryBuffer buffer = getNextAvailableBuffer();
+
+                // Write the data to the buffer.
+                BufferView slice = data.slice(data.getLength() - remainingLength, remainingLength);
+                DirectMemoryBuffer.WriteResult writeResult = buffer.write(slice, lastBlockAddress);
+                if (writeResult == null) {
+                    // Someone else grabbed this buffer and wrote to it before we got a chance. Go back and find another one.
+                    continue;
+                }
+
+                // Update our state, including the number of bytes written. In case of a subsequent error (and rollback),
+                // invoking delete() will undo this changes as well.
+                assert writeResult.getWrittenLength() >= 0 && writeResult.getWrittenLength() <= remainingLength : writeResult.getWrittenLength();
+                remainingLength -= writeResult.getWrittenLength();
+                this.storedBytes.addAndGet(writeResult.getWrittenLength());
+                lastBlockAddress = writeResult.getLastBlockAddress();
+            }
+        } catch (Throwable ex) {
+            if (!Exceptions.mustRethrow(ex) && lastBlockAddress != CacheLayout.NO_ADDRESS) {
+                // We wrote something, but got interrupted. We need to clean up whatever we wrote so we don't leave
+                // unreferenced data in the cache.
+                delete(lastBlockAddress);
+            }
+            throw ex;
+        }
+
+        CacheMetrics.insert(data.getLength());
+        return lastBlockAddress;
+    }
+
+    @Override
+    public int replace(int address, BufferView data) {
+        // The simplest and safest way to do a replace is to first insert the new data and then remove the old one. This
+        // will ensure the Cache remains consistent during the operation and we won't need to worry about concurrency or
+        // potential CacheFullExceptions.
+        int newAddress = insert(data);
+        delete(address);
+        return newAddress;
+    }
+
+    @Override
+    public int getAppendableLength(int currentLength) {
+        // If the current length is 0, we still allocate a block for it, so we need to indicate we can return the full
+        // block length for that. Otherwise we can only append up to the end of the last block.
+        int lastBlockLength = currentLength % this.layout.blockSize();
+        return currentLength == 0 ? this.layout.blockSize() : (lastBlockLength == 0 ? 0 : this.layout.blockSize() - lastBlockLength);
+    }
+
+    @Override
+    public int append(int address, int expectedLength, BufferView data) {
+        Exceptions.checkNotClosed(this.closed.get(), this);
+        Preconditions.checkArgument(address != CacheLayout.NO_ADDRESS, "Invalid address.");
+
+        // We can only append to fill the last block. For anything else a new write will be needed.
```

Comment:
According to this comment and the code, if the append data overflows the last block, then a new write will be needed. I'm not sure what this implies, is it that the whole entry needs to be reinserted?

Reply:
No. See https://github.com/pravega/pravega/pull/4074#discussion_r353985267.

Appends have been added especially to support streaming appends (1 byte, 10 bytes, 100 bytes, etc). They provide a very fast way of quickly adding that data to an existing entry without the need to allocate new blocks.

If we need to append more than 4KB (or whatever is available), the `ReadIndex` will fill up the last block and then create a new entry for the remaining data (for which it will also add a pointer in its index).